### PR TITLE
QT CIA installation

### PR DESCRIPTION
**Pull request by BreadFish64**
Adds QT CIA installation
image
image
image

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3144)

### CODE REVIEW

File: src/citra_qt/main.cpp [hunk 1]
```diff
@@ -696,6 +702,64 @@ void GMainWindow::OnMenuSelectGameListRoot() {
...
+        this, tr("Load File"), UISettings::values.roms_path,
+        tr("3DS Installation File (*.CIA*)") + ";;" + tr("All Files (*.*)"));
+    if (filepath.isEmpty())
+        return;
+
+    ui.action_Install_CIA->setEnabled(false);
+
+    // update progress
+    connect(this, &GMainWindow::UpdateProgress, this, &GMainWindow::OnUpdateProgress);
+    watcher = new QFutureWatcher<Service::AM::InstallStatus>;
```
Comment (MerryMage) [inline src/citra_qt/main.cpp]: Memory leak.

File: src/citra_qt/main.cpp [hunk 2]
```diff
@@ -696,6 +702,64 @@ void GMainWindow::OnMenuSelectGameListRoot() {
...
+    QString filepath = QFileDialog::getOpenFileName(
+        this, tr("Load File"), UISettings::values.roms_path,
+        tr("3DS Installation File (*.CIA*)") + ";;" + tr("All Files (*.*)"));
+    if (filepath.isEmpty())
+        return;
+
+    ui.action_Install_CIA->setEnabled(false);
+
+    // update progress
+    connect(this, &GMainWindow::UpdateProgress, this, &GMainWindow::OnUpdateProgress);
```
Comment (MerryMage) [inline src/citra_qt/main.cpp]: On repeated invocations of `OnMenuInstallCIA`, `connect` will be called multiple times, resulting in multiple calls to `OnUpdateProgress` on each signal emission.

On Qt 4.6 the lazy fix is to use `Qt::UniqueConnection`. The proper fix is to put this in the constructor.

File: src/citra_qt/main.cpp [hunk 3]
```diff
@@ -696,6 +702,64 @@ void GMainWindow::OnMenuSelectGameListRoot() {
...
+    watcher->setFuture(f);
+}
+
+void GMainWindow::OnUpdateProgress(size_t written, size_t total) {
+    progress_bar->setMaximum(total);
+    progress_bar->setValue(written);
+}
+
+void GMainWindow::OnCIAInstallFinished() {
+    this->statusBar()->removeWidget(progress_bar);
```
Comment (MerryMage) [inline src/citra_qt/main.cpp]: Memory leak of `progress_bar`.

File: src/citra_qt/main.cpp (hunk 2 above)
Comment (BreadFish64) [reply src/citra_qt/main.cpp]: moved to ConnectWidgetEvents

File: src/citra_qt/main.cpp (hunk 1 above)
Comment (BreadFish64) [reply src/citra_qt/main.cpp]: fixed (I think)

File: src/citra_qt/main.cpp (hunk 3 above)
Comment (BreadFish64) [reply src/citra_qt/main.cpp]: fixed

**Comment by B3n30** [comment]
This will most likely conflict with #3073. So be aware of that before adding it to canary or merging it

File: src/citra_qt/main.h [hunk 4]
```diff
@@ -149,8 +156,11 @@ private slots:
 
     GRenderWindow* render_window;
     GameList* game_list;
+    QFutureWatcher<Service::AM::InstallStatus>* watcher = nullptr;
 
     // Status bar elements
+
```
Comment (B3n30) [inline src/citra_qt/main.h]: remove the new line

File: src/citra_qt/main.cpp [hunk 5]
```diff
@@ -696,6 +708,61 @@ void GMainWindow::OnMenuSelectGameListRoot() {
     }
 }
 
+void GMainWindow::OnMenuInstallCIA() {
+    QString filepath = QFileDialog::getOpenFileName(
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: It would be nice to have getOpenFileNames and be able to install multiple files at once. But I know it is a bit more difficult for error handling and the progress bar...
So maybe in the future(tm)?

Comment (BreadFish64) [reply src/citra_qt/main.cpp]: Yeah, I specifically disable the button while installing the CIA to avoid that nightmare

File: src/citra_qt/main.cpp [hunk 6]
```diff
@@ -878,9 +945,12 @@ void GMainWindow::OnCoreError(Core::System::ResultStatus result, std::string det
             this, tr("Fatal Error"),
             tr("Citra has encountered a fatal error, please see the log for more details. "
                "For more information on accessing the log, please see the following page: "
-               "<a href='https://community.citra-emu.org/t/how-to-upload-the-log-file/296'>How to "
-               "Upload the Log File</a>.<br/><br/>Would you like to quit back to the game list? "
-               "Continuing emulation may result in crashes, corrupted save data, or other bugs."),
+               "<a href='https://community.citra-emu.org/t/how-to-upload-the-log-file/296'>How "
```
[the remark this answers is not in the document]
Comment (BreadFish64) [reply src/citra_qt/main.cpp]: Good question... I suspect clang has something to do with it

[the remark this answers is not in the document]
Comment (BreadFish64) [reply src/citra_qt/main.cpp]: fixed

File: src/citra_qt/main.h [hunk 7]
```diff
@@ -5,9 +5,12 @@
 #pragma once
 
 #include <memory>
+#include <QFutureWatcher>
```
[the remark this answers is not in the document]
Comment (BreadFish64) [reply src/citra_qt/main.h]: fixed

File: src/citra_qt/main.h (hunk 4 above)
Comment (BreadFish64) [reply src/citra_qt/main.h]: fixed

File: src/citra_qt/main.h [hunk 8]
```diff
@@ -5,30 +5,31 @@
...
 #include <QMainWindow>
 #include <QTimer>
-#include <QtWidgets/qprogressbar.h>
 #include "core/core.h"
 #include "core/hle/service/am/am.h"
 #include "ui_main.h"
 
+class AboutDialog;
 class Config;
+class QProgressBar;
```
Comment (B3n30) [inline src/citra_qt/main.h]: Since the sorting here is alphabetical this should be below class QFutureWatcher

File: src/citra_qt/main.h [hunk 9]
```diff
@@ -5,30 +5,31 @@
...
 class EmuThread;
 class GameList;
 class GImageInfo;
-class GPUCommandStreamWidget;
 class GPUCommandListWidget;
+class GPUCommandStreamWidget;
```
Comment (B3n30) [inline src/citra_qt/main.h]: Why did you change this?

File: src/citra_qt/main.h [hunk 10]
```diff
@@ -8,24 +8,28 @@
 #include <QMainWindow>
 #include <QTimer>
 #include "core/core.h"
+#include "core/hle/service/am/am.h"
 #include "ui_main.h"
 
+class AboutDialog;
```
Comment (B3n30) [inline src/citra_qt/main.h]: Unintentional Change? But it's fine

**Comment by Subv** [comment]
Please squash this
